src/main/java/org/jenkinsci/main/modules/sshd/AsynchronousCommand.java

```diff
@@ -112,20 +109,16 @@ public void start(Environment env) throws IOException {
     public void run() {
         try {
             int i;
-
-            // run the command in the context of the authenticated user
-            SecurityContext old = SecurityContextHolder.getContext();
             User user = getCurrentUser();
-            if (user!=null)
-                ACL.impersonate(user.impersonate());
-
-            try {
+            if(user!=null){
+              try(ACLContext ctx = ACL.as(user)){
                 i = AsynchronousCommand.this.runCommand();
-            } finally {
-                out.flush(); // working around SSHD-154
-                err.flush();
-                SecurityContextHolder.setContext(old);
+              }
+            } else {
+              i = AsynchronousCommand.this.runCommand();
             }
+            out.flush(); // working around SSHD-154
+            err.flush();
```

Comment:
Do we still want these in a `finally` block for diagnostic purposes? Maybe around line 131?

Reply:
Due to https://issues.apache.org/jira/browse/SSHD-154 is resolved I think we can rid of those flush, or copy them to the catch block too just in case, in any case, the finally block does not make sense as it was.

Reply:
I have copy them to the catch block